File: solver/greedy/greedy.py
from state.state import State
from direction.direction import Direction
from solver.solver import Solver


class GreedySolver(Solver):
    def solve(self, initial: State):
        increment = 1
        visited: set[tuple[int, ...]] = set()
        heap: list[tuple[int, int, State]] = []
        Solver._push(heap, 0, (increment := increment + 1), initial)
        while len(heap) != 0:
            _, _, state = Solver._pop(heap)
            if state.matrix == self.goal.matrix:
                print("found,", len(visited), 'states expanded,',
                      state.depth, 'depths')
                state.print_trace()
                break
            for i in Direction:
                moved = state.move(i)
                if moved is None or tuple(moved.matrix) in visited:
                    continue
                Solver._push(heap, Solver._heuristics(
                    moved, self.goal), (increment := increment + 1), moved)
                Solver._visit(visited, moved.matrix)

File: src/solver/greedy/greedy.cpp
#include "greedy.hpp"
#include "../../direction/direction.hpp"
#include "../../heap/heap.hpp"
#include <algorithm>
#include <chrono>
#include <iostream>
#include <list>
#include <thread>
#include <unordered_set>

namespace {
struct BoardEqual {
    bool operator()(const Board* a, const Board* b) const {
        return a->equals(b);
    }
};
} // namespace

using namespace std::chrono_literals;
const Solver* GreedySolver::solve(const Board* initial) const {
    auto heap = new Heap();
    std::unordered_set<const Board*, std::hash<const Board*>, BoardEqual>
        visited;
    heap->push(new Node(
        std::make_pair((Board*)initial, initial->getManhattanDistance(goal))));
    while (heap->length()) {
        auto node = heap->pop();
        auto board = node->value().first;
        visited.emplace(board);
        if (board->equals(goal)) {
            board->printTrace();
            std::cout << "Found, " << visited.size() << " states explored."
                      << std::endl;
            break;
        }
        for (int i = 0; i != 4; ++i) {
            auto moved = board->move((Direction)(i)).value_or(nullptr);
            if (moved == nullptr || visited.count(moved)) {
                continue;
            }
            heap->push(new Node(
                std::make_pair(moved, node->value().second +
                                          moved->getManhattanDistance(goal))));
            visited.emplace(moved);
        }
    }
    return this;
}
